### DISCUSSION TITLE
Batch apply -composite with bounding box input from "%@"

### DISCUSSION THREAD

**Discussion by carlosnasillo** [ImageMagick/ImageMagick#1975] [General] [answered]
I have the following two transparent images which I'd like to overlay

`yamaha.png`
![Bike](https://i.imgur.com/MtSMAc9.png)

`shadow.png`

![Shadow](https://i.imgur.com/GuOxCa3.png)

`result.png`
![Result Manual](https://i.imgur.com/duwXZJh.png)

I am currently able to do it manually by first running 

`convert yamaha.png -fuzz 1% -format "%@" info:`

> 450x307+52+124

And then I take the last two values for the `-geometry` within the `-composite` command.
But I don't use the raw value, I manually subtract `+124` from the bike's image height `x555`, resulting in `+431` in order to place the shadow at the bottom of the bounding box.

Therefore the final command is

`convert yamaha.png shadow.png -geometry +52+431 -composite result.png`

This works well and does the job, however: 

**Question**

How can I do (1) retrieval of the bounding box and pass it as an input, (2) perform a subtraction on the geometry parameter, (3) run the command in batch mode for all items in a folder? 

Many thanks!

**Comment by fmw42** [comment]
What is your IM version and platform?  That can be done in IM 7, but probably not in IM 6.

**Reply by carlosnasillo** [reply]
```
Version: ImageMagick 7.0.10-10 Q16 x86_64 2020-04-28 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1) 
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms ltdl lzma openexr png ps tiff webp xml zlib
```

**Comment by fmw42** [comment, marked as answer]
Actually it is easier just to append the two images after doing a -trim.

`magick bike.png -fuzz 1% -trim +repage shadow.png -gravity south -background white -smush 0 result.png`

![result](https://user-images.githubusercontent.com/13458218/80924204-7f35ae00-8d3c-11ea-9526-1e0a76a08db2.png)

If you want more or less space between the two images, change the argument to -smush from 0 to a positive or negative value, respectively.

Also you can either add a border of white or use -extent if you want the result to have more white space around the whole image.

**Reply by carlosnasillo** [reply]
Amazing thanks! But how can I do it in batch? I tried `mogrify` with `-path` but fails. 

`mogrify -path . *.png -fuzz 1% -trim +repage shadow.png -gravity south -background white -smush -7`
> Abort trap: 6

I've never used the `magick` command... always either `convert` or `mogrify`

**Comment by fmw42** [comment]
Mogrify uses -draw to do compositing. Otherwise, you cannot have a second image in the command line. But mogrify is IM 6 technology and does not allow in-line computations such as %@. So the best recommendation is to write a script loop over your images and your shadow(s).  That would be OS dependent and you did not say which platform you are using. How many shadows do you have? Is it one shadow for all bike images pairs -- shadow and bike image are both different?

magick is the replacement for convert in IM 7.  For mogrify and all other tools, you need to preface magick before mogrify, etc. The syntax is generally the same. There are a few differences. See https://imagemagick.org/script/porting.php#cli

**Reply by carlosnasillo** [reply]
Ok thank you for the explanation. New to ImageMagick.
Im running MacOS 10.15.4
I have about 5 shadow variations, which I hoped to randomly pick one for any one bike.

**Comment by fmw42** [comment]
When you say randomly pick, do you mean you want that to be automatic or are you picking them manually?

**Reply by carlosnasillo** [reply]
Apologies. I just meant to convey that I have 5 shadow image files and do not have a preference which of these get applied for `+repage`. 

In my mind I imagined something along the lines of:

`X=$(awk -v min=1 -v max=5 'BEGIN{srand(); print int(min+rand()*(max-min+1))}')`
`magick bike.png -fuzz 1% -trim +repage shadow_$X.png ...` 

Therefore; Yes, automatic.

**Comment by fmw42** [comment]
So you can script it as you like as it looks like you know bash scripting for the Mac.  But note that ImageMagick also has a random() function in -fx in case you want to use that.

**Reply by carlosnasillo** [reply]
Indeed. I was just wondering if there was an out-of-the-box command. 
Perfect, thanks again for all your help. 

BTW Just checked `fmwconcepts.com/imagemagick/` - kudos, its amazing work.
